### ISSUE TITLE
newsboat broken in single-user mode

### ISSUE THREAD

**Issue by yurkobb** [NixOS/nixpkgs#60845]
## Issue description
This is a consequence of NixOS/nix#2762.

### Steps to reproduce
```
$ nix-env -i newsboat
replacing old 'newsboat-2.14.1'
installing 'newsboat-2.15'
these derivations will be built:
  /nix/store/3krczr2yrnpnjqjaw23a11k543ywnsza-newsboat-2.15.drv
building '/nix/store/3krczr2yrnpnjqjaw23a11k543ywnsza-newsboat-2.15.drv'...
unpacking sources
unpacking source archive /nix/store/ndqhrldc89207rqi8mda8nz6lj4vdpq9-newsboat-2.15.tar.xz
do not know how to unpack source archive /nix/store/ndqhrldc89207rqi8mda8nz6lj4vdpq9-newsboat-2.15.tar.xz
builder for '/nix/store/3krczr2yrnpnjqjaw23a11k543ywnsza-newsboat-2.15.drv' failed with exit code 1
error: build of '/nix/store/3krczr2yrnpnjqjaw23a11k543ywnsza-newsboat-2.15.drv' failed
```

## Technical details

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the
results.

```
 - system: `"x86_64-linux"`
 - host os: `Linux 5.0.9-200.fc29.x86_64, Fedora, 29 (Twenty Nine)`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.2`
 - channels(yurb): `"nixpkgs-19.09pre178281.76fad451105"`
 - nixpkgs: `/home/yurb/.nix-defexpr/channels/nixpkgs`
```

cc @nicknovitski @dotlambda

**Comment by nicknovitski** [comment]
I'm able to run install it correctly with a single-user set-up.

```
$ nix-env -iA nixpkgs.newsboat
installing 'newsboat-2.16.1'
these paths will be fetched (4.12 MiB download, 29.65 MiB unpacked):
  /nix/store/1qdhqhpn6cggqa7yka0p4kka372pl0s3-json-c-0.13.1
  /nix/store/5dz4q2prci96hqsp79gn4zrmzyzsj7i7-stfl-0.24
  /nix/store/nvg2h34h3987j7rv9lvrp8isr1gj6jdz-newsboat-2.16.1
copying path '/nix/store/1qdhqhpn6cggqa7yka0p4kka372pl0s3-json-c-0.13.1' from 'https://cache.nixos.org'...
copying path '/nix/store/5dz4q2prci96hqsp79gn4zrmzyzsj7i7-stfl-0.24' from 'https://cache.nixos.org'...
copying path '/nix/store/nvg2h34h3987j7rv9lvrp8isr1gj6jdz-newsboat-2.16.1' from 'https://cache.nixos.org'...
building '/nix/store/a6al6xna48xwlnbqzwy8alf2fmcllpf3-user-environment.drv'...
created 1496 symlinks in user environment
$ nix-shell -p nix-info --run "nix-info -m"
these paths will be fetched (0.01 MiB download, 0.03 MiB unpacked):
  /nix/store/5bl702li6rpdvsp6p9dh0gk9cgvk6fry-DarwinTools-1
  /nix/store/hbcf4i5qr3lx5l4sy1wdrjmm0b87p98s-nix-info
copying path '/nix/store/5bl702li6rpdvsp6p9dh0gk9cgvk6fry-DarwinTools-1' from 'https://cache.nixos.org'...
copying path '/nix/store/hbcf4i5qr3lx5l4sy1wdrjmm0b87p98s-nix-info' from 'https://cache.nixos.org'...
 - system: `"x86_64-darwin"`
 - host os: `Darwin 18.6.0, macOS 10.14.5`
 - multi-user?: `no`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.2.2`
 - channels(nick): `"home-manager, nixpkgs-19.09pre186107.6c0ed987eed"`
 - nixpkgs: `/Users/nick/.nix-defexpr/channels/nixpkgs`
```
Unlike you I do have sandboxing disabled, because of too many packages that fail when they (i think) impurely access macos frameworks.

**Comment by yurkobb** [comment]
That's interesting - I didn't actually enable sandbox, and I thought it is not possible to use if nix does not have root privileges ([per manual](https://nixos.org/nix/manual/#sec-conf-file)).

Anyway, I tried explicitly disabling it both by adding `--option sandbox false` on command line and adding `sandbox = false` to `~/.config/nix/nix.conf`, yet `nix-info` still reports it as being enabled for some reason, and newsboat doesn't build.

So there are two questions here:
- why `nix-info` reports sandbox as `yes` if I don't have it enabled and/or have overridden it on command line (`nix-shell --option sandbox false -p nix-info --run "nix-info -m"`.)
- how else can I verify it is/isn't using sandbox to build newsboat?

**Comment by yurkobb** [comment]
Ah, can you try also without using cache? From the `nix-env` output you posted it seems like it just used a cached build from the server.

**Comment by nicknovitski** [comment]
Very true.  I tried editing the derivation and confirmed I was able to build it from source, obviously unpacking the source in the process.

Another thing that might make a difference is that I'm using the darwin stdenv.

**Comment by dywedir** [comment]
Minimal repro:

```
{ stdenv, fetchurl }:

stdenv.mkDerivation rec {
  name = "newsboat-${version}";
  version = "2.15";

  src = fetchurl {
    url = "https://newsboat.org/releases/${version}/${name}.tar.xz";
    sha256 = "1dqdcp34jmphqf3d8ik0xdhg0s66nd5rky0y8y591nidq29wws6s";
  };

  unpackPhase = ''
    xz -d < "$src" | tar xf -
  '';
}
```

```
builder for '/nix/store/vz71rdakp2px7jf49s2id8vr7pb21rns-newsboat-2.15.drv' failed with exit code 141;
```

```
> tar -tJf newsboat-2.16.1.tar.xz -i --block-number | nl | tail -n 25
/run/current-system/sw/bin/tar: Skipping to next header
/run/current-system/sw/bin/tar: Exiting with failure status due to previous errors
   402	block 7855: newsboat-2.16.1/test/rssitem.cpp
   403	block 7858: newsboat-2.16.1/test/rsspp_parser.cpp
   404	block 7873: newsboat-2.16.1/test/rsspp_rssparser.cpp
   405	block 7879: newsboat-2.16.1/test/strprintf.cpp
   406	block 7896: newsboat-2.16.1/test/tagsouppullparser.cpp
   407	block 7903: newsboat-2.16.1/test/test-helpers.cpp
   408	block 7920: newsboat-2.16.1/test/test-helpers.h
   409	block 7946: newsboat-2.16.1/test/test.cpp
   410	block 7948: newsboat-2.16.1/test/textformatter.cpp
   411	block 7962: newsboat-2.16.1/test/utils.cpp
   412	block 8025: newsboat-2.16.1/txt2h
   413	block 8027: ** Block of NULs **
   414	block 8028: ** Block of NULs **
   415	block 8029: ** Block of NULs **
   416	block 8030: ** Block of NULs **
   417	block 8031: ** Block of NULs **
   418	block 8032: ** Block of NULs **
   419	block 8033: ** Block of NULs **
   420	block 8034: ** Block of NULs **
   421	block 8035: ** Block of NULs **
   422	block 8036: ** Block of NULs **
   423	block 8037: ** Block of NULs **
   424	block 8038: ** Block of NULs **
   425	block 8039: ** Block of NULs **
   426	block 8041: block 8077: ** End of File **
```

**Comment by yurkobb** [comment]
xz or tar misbehaving on x86_64-linux?

**Comment by timokau** [comment]
The same thing happens on [hydra](https://hydra.nixos.org/job/nixos/trunk-combined/nixpkgs.newsboat.x86_64-linux). The build does occasionally succeed (but rarely). It always succeeds locally for me. Sandboxing is enabled.

May be related to https://github.com/NixOS/nixpkgs/issues/20950.

**Comment by Minoru** [comment]
For what it's worth, note that Newsboat's tarballs are packed with [pixz](https://github.com/vasi/pixz), which is an *indexed* compressor. It seems compatible with original `xz`—I just checked, and xz 5.2.4 decompressed pixz 1.0.6 output just fine—but perhaps some versions are indeed incompatible?

This won't explain sporadic failures, though.

**Comment by timokau** [comment]
Yeah the sporadic failures really shouldn't be happening.

As long as we don't know what actually causes this, https://github.com/NixOS/nixpkgs/pull/76175 may work around it.

**Comment by timokau** [comment]
The workaround [worked](https://hydra.nixos.org/job/nixos/trunk-combined/nixpkgs.newsboat.x86_64-linux). However I'm unsure if we should close this, since it was not the only problem of this kind and we still don't have any clue what the cause was.

**Comment by vcunat** [comment]
The SIGPIPE should now be addressed by PR #286721
